File: java-kata/src/main/java/com/leetcode/MaximumDepthOfBinaryTree.java
package com.leetcode;

import com.leetcode.data.TreeNode;

/**
 * <h3><a href="https://leetcode.com/problems/maximum-depth-of-binary-tree/">Maximum Depth of Binary Tree</a></h3>
 * <p>
 *     Given the <code>root</code> of a binary tree, return its <em>maximum depth</em>.
 * </p>
 * <p>
 *     A binary tree's <strong>maximum depth</strong> is the number of nodes along the longest path
 *     from the root node down to the farthest leaf node.
 * </p>
 * <p>
 *     <h6>Constraints:</h6>
 *     <ul>
 *         <li>The number of nodes in the tree is in the range <code>[0, 10^4]</code></li>
 *         <li><code>-100 <= Node.val <= 100</code></li>
 *     </ul>
 * </p>
 */
public class MaximumDepthOfBinaryTree {

    public int maxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        if (root.left == null && root.right == null) {
            return 1;
        }
        return 1 + Math.max(maxDepth(root.left), maxDepth(root.right));
    }
}

File: kotlin-kata/src/main/kotlin/com/leetcode/MaximumDepthOfBinaryTree.kt
package com.leetcode

import com.leetcode.data.TreeNode
import kotlin.math.max

/**
 * ### [Maximum Depth of Binary Tree](https://leetcode.com/problems/maximum-depth-of-binary-tree/)
 *
 * Given the `root` of a binary tree, return its *maximum depth*.
 *
 * A binary tree's **maximum depth** is the number of nodes along the longest path
 * from the root node down to the farthest leaf node.
 *
 * ###### Constraints:
 *  - The number of nodes in the tree is in the range `[0, 10^4]`
 *  - `-100 <= Node.val <= 100`
 */
class MaximumDepthOfBinaryTree {
    fun maxDepth(root: TreeNode?): Int {
        if (root == null) {
            return 0
        }
        if (root.left == null && root.right == null) {
            return 1
        }
        return 1 + max(maxDepth(root.left), maxDepth(root.right))
    }
}
